### PR TITLE
Issue/5316 – Generic KeyValueStorage

### LINKED ISSUE #5316: Create generic KeyValueStorage for testable NSUserDefaults persistence
Based on [this discussion](https://github.com/wordpress-mobile/WordPress-iOS/pull/5232#discussion_r62207322) with @koke, let's make `MediaSettingsStorage` into a generic `KeyValueStorage`. This approach will allow us to remove `clearUserDefaults` from `StoreCoordinatorTests`.

### PR DESCRIPTION
**Pull request by frosty**
Fixes #5316.

I've extracted `KeyValueStorage` from `MediaSettings`, and used it within `StoreCoordinator` (instead of directly using `NSUserDefaults`). This means that StoreCoordinatorTests no longer need to manually clear specific defaults keys – they can use an InMemoryDatabase instead.

Needs review: @koke 
Thanks!

### CODE REVIEW

File: WordPress/Classes/Services/KeyValueStorage.swift [hunk 1]
```diff
@@ -0,0 +1,43 @@
+import Foundation
+
+protocol KeyValueStorage {
+    func valueForKey(key: String) -> AnyObject?
+    func setValue(value: AnyObject, forKey key: String)
```
Comment (koke) [inline WordPress/Classes/Services/KeyValueStorage.swift]: Looking at `NSUserDefaults.setObject(_:forKey:)` documentation:

> The value parameter can be only property list objects: NSData, NSString, NSNumber, NSDate, NSArray, or NSDictionary. For NSArray and NSDictionary objects, their contents must be property list objects. See What is a Property List? in Property List Programming Guide.

Considering the expected usage, I don't think it's worth trying to serialize other types, but the protocol documentation should list the supported types.

**Comment by koke** [comment]
I'm not 100% convinced, but here's my suggestion around naming:

```swift
protocol KeyValueDatabase {}
struct DefaultsKeyValueDatabase: KeyValueDatabase {}
class EphemeralKeyValueDatabase: KeyValueDatabase {}
```

I know I suggested `Ephemeral` to `Memory` but I instantly regretted it 😄 

I'd also be OK with replacing every `Database` with `Storage`, but only one of them.

Also, what do you think about tweaking the API a bit so it matches `NSUserDefaults`?

```swift
protocol KeyValueDatabase {
    func objectForKey(key: String) -> AnyObject?
    func setObject(value: AnyObject?, forKey key: String)
    func removeObjectForKey(key: String)
}

extension NSUserDefaults: KeyValueDatabase {}

/// `InMemoryDatabase` stores values in a dictionary in memory, and is
/// never persisted between app launches.
class EphemeralKeyValueDatabase: KeyValueDatabase {
    private var memory = [String: AnyObject]()

    func setObject(value: AnyObject?, forKey key: String) {
        memory[key] = value
    }

    func objectForKey(key: String) -> AnyObject? {
        return memory[key]
    }

    func removeObjectForKey(key: String) {
        memory[key] = nil
    }
}
```

File: WordPress/Classes/Services/Store.swift [hunk 2]
```diff
@@ -48,33 +48,31 @@ enum StoreCoordinatorError: ErrorType {
 ///   purchased product, as well as a localized error message under `NSUnderlyingErrorKey`.
 class StoreCoordinator<S: Store> {
     private let store: S
+    private let keyValueStorage: KeyValueStorage
```
Comment (koke) [inline WordPress/Classes/Services/Store.swift]: I'd rename this, not sure to what. Naming it `keyValueStorage` feels like naming something `dictionary`. Maybe `settings` or `database`.

**Comment by frosty** [comment]
Thanks for the great review, @koke! 

Your naming suggestions look good. I actually did this whilst working on the PR, but I didn't follow through on it: `extension NSUserDefaults: KeyValueDatabase {}`. It's a good idea though, so I've updated the `KeyValueDatabase` protocol to match `NSUserDefaults.

All other changes implemented too – ready for another review!

**Comment by koke** [comment]
Looks good, :shipit: once Travis goes green

**Pull request merged**